File: go-sidecar/signer/status_list_test.go
// Tests for the W3C BitstringStatusList implementation (W3C CG Report §11.2).
//
// Mirrors tests/test_status_list.py and packages/sdk-ts/tests/status-list.test.ts.
// Cross-language interop is verified against the canonical test vector at
// test-vectors/bitstring-status-list/vector.json.

package signer

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"testing"
)

const (
	testStatusURL  = "https://issuer.example/status/1"
	testIssuerDID  = "did:web:issuer.example"
)

// ---------------------------------------------------------------------------
// Construction & validation
// ---------------------------------------------------------------------------

func TestStatusListDefaultLengthAndPurpose(t *testing.T) {
	sl, err := NewStatusList(testStatusURL, "", 0)
	if err != nil {
		t.Fatal(err)
	}
	if sl.Length() != DefaultBitstringLength {
		t.Errorf("Length = %d, want %d", sl.Length(), DefaultBitstringLength)
	}
	if sl.StatusPurpose() != StatusPurposeRevocation {
		t.Errorf("StatusPurpose = %q, want %q", sl.StatusPurpose(), StatusPurposeRevocation)
	}
}

func TestStatusListRejectsShortBitstring(t *testing.T) {
	if _, err := NewStatusList(testStatusURL, StatusPurposeRevocation, 1024); err == nil {
		t.Error("expected error for short length")
	}
}

func TestStatusListRejectsNonMultipleOfEight(t *testing.T) {
	if _, err := NewStatusList(testStatusURL, StatusPurposeRevocation, DefaultBitstringLength+1); err == nil {
		t.Error("expected error for non-multiple-of-8 length")
	}
}

func TestStatusListRejectsInvalidPurpose(t *testing.T) {
	if _, err := NewStatusList(testStatusURL, "bogus", 0); err == nil {
		t.Error("expected error for invalid purpose")
	}
}

func TestStatusListRejectsEmptyID(t *testing.T) {
	if _, err := NewStatusList("", StatusPurposeRevocation, 0); err == nil {
		t.Error("expected error for empty id")
	}
}

func TestStatusListDefaultStateAllZero(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	for _, idx := range []int{0, 1, 7, 8, 100, 65535, DefaultBitstringLength - 1} {
		got, err := sl.GetStatus(idx)
		if err != nil {
			t.Fatal(err)
		}
		if got {
			t.Errorf("GetStatus(%d) = true, want false", idx)
		}
	}
}

// ---------------------------------------------------------------------------
// Bit operations
// ---------------------------------------------------------------------------

func TestSetAndGetStatus(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	if err := sl.SetStatus(42, true); err != nil {
		t.Fatal(err)
	}

	if got, _ := sl.GetStatus(42); !got {
		t.Error("expected bit 42 to be set")
	}
	if got, _ := sl.GetStatus(41); got {
		t.Error("expected bit 41 to be unset")
	}
	if got, _ := sl.GetStatus(43); got {
		t.Error("expected bit 43 to be unset")
	}
}

func TestClearAfterSet(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	_ = sl.SetStatus(42, true)
	_ = sl.SetStatus(42, false)
	if got, _ := sl.GetStatus(42); got {
		t.Error("expected bit 42 to be cleared")
	}
}

func TestFirstAndLastBit(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	_ = sl.Revoke(0)
	_ = sl.Revoke(DefaultBitstringLength - 1)
	if got, _ := sl.IsSet(0); !got {
		t.Error("expected first bit to be set")
	}
	if got, _ := sl.IsSet(DefaultBitstringLength - 1); !got {
		t.Error("expected last bit to be set")
	}
	if got, _ := sl.IsSet(1); got {
		t.Error("expected bit 1 to be unset")
	}
}

func TestOutOfRangeIndex(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	if err := sl.SetStatus(-1, true); err == nil {
		t.Error("expected error for negative index")
	}
	if err := sl.SetStatus(DefaultBitstringLength, true); err == nil {
		t.Error("expected error for out-of-range index")
	}
}

// ---------------------------------------------------------------------------
// Allocation
// ---------------------------------------------------------------------------

func TestAllocateIndicesAreSequential(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	for want := 0; want < 3; want++ {
		got, err := sl.AllocateIndex()
		if err != nil {
			t.Fatal(err)
		}
		if got != want {
			t.Errorf("AllocateIndex = %d, want %d", got, want)
		}
	}
}

// ---------------------------------------------------------------------------
// Encoding
// ---------------------------------------------------------------------------

func TestEncodeUsesMultibasePrefix(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	encoded, err := sl.Encode()
	if err != nil {
		t.Fatal(err)
	}
	if encoded[:1] != MultibaseBase64URLPrefix {
		t.Errorf("encoded prefix = %q, want %q", encoded[:1], MultibaseBase64URLPrefix)
	}
}

func TestEncodeRoundTrip(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	indices := []int{0, 1, 7, 8, 9, 16, 1023, 65535, DefaultBitstringLength - 1}
	for _, idx := range indices {
		_ = sl.Revoke(idx)
	}
	encoded, err := sl.Encode()
	if err != nil {
		t.Fatal(err)
	}
	decoded, err := DecodeStatusList(encoded, testStatusURL, StatusPurposeRevocation)
	if err != nil {
		t.Fatal(err)
	}
	for _, idx := range indices {
		got, _ := decoded.GetStatus(idx)
		if !got {
			t.Errorf("decoded.GetStatus(%d) = false, want true", idx)
		}
	}
	if got, _ := decoded.GetStatus(2); got {
		t.Error("expected bit 2 to be unset after roundtrip")
	}
}

func TestDecodeRejectsWrongPrefix(t *testing.T) {
	if _, err := DecodeStatusList("zSomethingBase58", testStatusURL, ""); err == nil {
		t.Error("expected error for wrong prefix")
	}
}

func TestDecodeRejectsCorruptPayload(t *testing.T) {
	if _, err := DecodeStatusList(MultibaseBase64URLPrefix+"$$$invalid$$$", testStatusURL, ""); err == nil {
		t.Error("expected error for corrupt payload")
	}
}

// ---------------------------------------------------------------------------
// Credential and entry builders
// ---------------------------------------------------------------------------

func TestBuildStatusListCredentialShape(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	_ = sl.Revoke(7)
	vc, err := BuildStatusListCredential(BuildStatusListCredentialOptions{
		IssuerDID:  testIssuerDID,
		StatusList: sl,
	})
	if err != nil {
		t.Fatal(err)
	}
	if vc["id"] != testStatusURL {
		t.Errorf("id = %v, want %s", vc["id"], testStatusURL)
	}
	if vc["issuer"] != testIssuerDID {
		t.Errorf("issuer = %v, want %s", vc["issuer"], testIssuerDID)
	}
	typeField, _ := vc["type"].([]any)
	if !containsString(typeField, BitstringStatusListCredentialType) {
		t.Errorf("type does not contain %s", BitstringStatusListCredentialType)
	}
	subject, _ := vc["credentialSubject"].(map[string]any)
	if subject["statusPurpose"] != StatusPurposeRevocation {
		t.Errorf("statusPurpose = %v, want %s", subject["statusPurpose"], StatusPurposeRevocation)
	}
	encoded, _ := subject["encodedList"].(string)
	if encoded[:1] != MultibaseBase64URLPrefix {
		t.Errorf("encodedList prefix = %q, want %q", encoded[:1], MultibaseBase64URLPrefix)
	}
}

func TestBuildStatusListEntryShape(t *testing.T) {
	entry, err := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: testStatusURL,
		StatusListIndex:      42,
	})
	if err != nil {
		t.Fatal(err)
	}
	if entry["statusListIndex"] != "42" {
		t.Errorf("statusListIndex = %v, want '42'", entry["statusListIndex"])
	}
	if entry["type"] != BitstringStatusListEntryType {
		t.Errorf("type = %v, want %s", entry["type"], BitstringStatusListEntryType)
	}
}

func TestBuildStatusListEntryRejectsNegativeIndex(t *testing.T) {
	_, err := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: testStatusURL,
		StatusListIndex:      -1,
	})
	if err == nil {
		t.Error("expected error for negative index")
	}
}

// ---------------------------------------------------------------------------
// VerifyStatus
// ---------------------------------------------------------------------------

func buildPair(t *testing.T, revoked ...int) (*StatusList, map[string]any) {
	t.Helper()
	sl, _ := NewStatusList(testStatusURL, "", 0)
	for _, idx := range revoked {
		_ = sl.Revoke(idx)
	}
	vc, err := BuildStatusListCredential(BuildStatusListCredentialOptions{
		IssuerDID:  testIssuerDID,
		StatusList: sl,
	})
	if err != nil {
		t.Fatal(err)
	}
	return sl, vc
}

func TestVerifyStatusUnsetReturnsFalse(t *testing.T) {
	_, vc := buildPair(t)
	entry, _ := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: testStatusURL,
		StatusListIndex:      10,
	})
	got, err := VerifyStatus(entry, vc)
	if err != nil {
		t.Fatal(err)
	}
	if got {
		t.Error("expected unset bit to return false")
	}
}

func TestVerifyStatusSetReturnsTrue(t *testing.T) {
	_, vc := buildPair(t, 10)
	entry, _ := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: testStatusURL,
		StatusListIndex:      10,
	})
	got, err := VerifyStatus(entry, vc)
	if err != nil {
		t.Fatal(err)
	}
	if !got {
		t.Error("expected set bit to return true")
	}
}

func TestVerifyStatusIDMismatchRaises(t *testing.T) {
	_, vc := buildPair(t)
	entry, _ := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: "https://other.example/status/9",
		StatusListIndex:      0,
	})
	if _, err := VerifyStatus(entry, vc); err == nil {
		t.Error("expected error for id mismatch")
	}
}

func TestVerifyStatusPurposeMismatchRaises(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, StatusPurposeSuspension, 0)
	vc, _ := BuildStatusListCredential(BuildStatusListCredentialOptions{
		IssuerDID:  testIssuerDID,
		StatusList: sl,
	})
	entry, _ := BuildStatusListEntry(BuildStatusListEntryOptions{
		StatusListCredential: testStatusURL,
		StatusListIndex:      0,
		StatusPurpose:        StatusPurposeRevocation,
	})
	if _, err := VerifyStatus(entry, vc); err == nil {
		t.Error("expected error for purpose mismatch")
	}
}

// ---------------------------------------------------------------------------
// Cross-language interop against the canonical test vector
// ---------------------------------------------------------------------------

type statusListVector struct {
	StatusListID                    string         `json:"status_list_id"`
	IssuerDID                       string         `json:"issuer_did"`
	BitstringLengthBits             int            `json:"bitstring_length_bits"`
	StatusPurpose                   string         `json:"status_purpose"`
	RevokedIndices                  []int          `json:"revoked_indices"`
	ActiveIndicesSample             []int          `json:"active_indices_sample"`
	ExpectedEncodedList             string         `json:"expected_encoded_list"`
	StatusListCredential            map[string]any `json:"status_list_credential"`
	SampleCredentialStatusRevoked   map[string]any `json:"sample_credential_status_revoked"`
	SampleCredentialStatusActive    map[string]any `json:"sample_credential_status_active"`
}

func loadStatusListVector(t *testing.T) statusListVector {
	t.Helper()
	wd, _ := os.Getwd()
	repoRoot := filepath.Clean(filepath.Join(wd, "..", ".."))
	vecPath := filepath.Join(
		repoRoot, "test-vectors", "bitstring-status-list", "vector.json",
	)
	raw, err := os.ReadFile(vecPath)
	if err != nil {
		t.Fatalf("read vector: %v", err)
	}
	var v statusListVector
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("decode vector: %v", err)
	}
	return v
}

func TestStatusListInteropEncodingDecodesEquivalent(t *testing.T) {
	// Go's compress/flate produces a valid DEFLATE stream that is byte-different
	// from Python/zlib's output (different DEFLATE encoder), but both decompress
	// to the same bitstring. W3C BitstringStatusList requires equivalence of the
	// decompressed bitstring, not the gzip envelope, so the right interop check
	// is decoded equivalence rather than encoded byte-identity.
	v := loadStatusListVector(t)
	sl, err := NewStatusList(v.StatusListID, v.StatusPurpose, v.BitstringLengthBits)
	if err != nil {
		t.Fatal(err)
	}
	for _, idx := range v.RevokedIndices {
		if err := sl.Revoke(idx); err != nil {
			t.Fatal(err)
		}
	}
	goEncoded, err := sl.Encode()
	if err != nil {
		t.Fatal(err)
	}

	goDecoded, err := DecodeStatusList(goEncoded, v.StatusListID, v.StatusPurpose)
	if err != nil {
		t.Fatal(err)
	}
	canonicalDecoded, err := DecodeStatusList(v.ExpectedEncodedList, v.StatusListID, v.StatusPurpose)
	if err != nil {
		t.Fatal(err)
	}

	if !bytesEqual(goDecoded.RawBytes(), canonicalDecoded.RawBytes()) {
		t.Errorf("Go-encoded bitstring does not decode equivalently to the canonical Python vector")
	}
}

func bytesEqual(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestStatusListInteropDecodesCanonical(t *testing.T) {
	v := loadStatusListVector(t)
	sl, err := DecodeStatusList(v.ExpectedEncodedList, v.StatusListID, v.StatusPurpose)
	if err != nil {
		t.Fatal(err)
	}
	for _, idx := range v.RevokedIndices {
		got, _ := sl.GetStatus(idx)
		if !got {
			t.Errorf("decoded.GetStatus(%d) = false, expected revoked", idx)
		}
	}
	for _, idx := range v.ActiveIndicesSample {
		got, _ := sl.GetStatus(idx)
		if got {
			t.Errorf("decoded.GetStatus(%d) = true, expected active", idx)
		}
	}
}

func TestStatusListInteropVerifyMatchesCanonical(t *testing.T) {
	v := loadStatusListVector(t)

	revoked, err := VerifyStatus(v.SampleCredentialStatusRevoked, v.StatusListCredential)
	if err != nil {
		t.Fatal(err)
	}
	if !revoked {
		t.Error("expected canonical revoked sample to verify as revoked")
	}

	active, err := VerifyStatus(v.SampleCredentialStatusActive, v.StatusListCredential)
	if err != nil {
		t.Fatal(err)
	}
	if active {
		t.Error("expected canonical active sample to verify as active")
	}
}

// Sanity-check that strconv.Itoa(0) yields "0" (used in entry builder).
func TestStringIndexFormat(t *testing.T) {
	if strconv.Itoa(0) != "0" {
		t.Error("strconv.Itoa(0) should yield '0'")
	}
}

// ---------------------------------------------------------------------------
// State dict persistence
// ---------------------------------------------------------------------------

func TestStateDictRoundTripPreservesBitsAndCursor(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	a, _ := sl.AllocateIndex()
	b, _ := sl.AllocateIndex()
	c, _ := sl.AllocateIndex()
	_ = sl.Revoke(b)

	state, err := sl.ToStateDict()
	if err != nil {
		t.Fatal(err)
	}

	restored, err := FromStateDict(state)
	if err != nil {
		t.Fatal(err)
	}
	if restored.StatusListID() != testStatusURL {
		t.Errorf("StatusListID mismatch: got %q", restored.StatusListID())
	}
	if got, _ := restored.GetStatus(a); got {
		t.Error("expected idx_a to be unset")
	}
	if got, _ := restored.GetStatus(b); !got {
		t.Error("expected idx_b to be set")
	}
	if got, _ := restored.GetStatus(c); got {
		t.Error("expected idx_c to be unset")
	}
	next, err := restored.AllocateIndex()
	if err != nil {
		t.Fatal(err)
	}
	if next != c+1 {
		t.Errorf("expected next allocation to be %d, got %d", c+1, next)
	}
}

func TestStateDictJSONSerializable(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	_ = sl.Revoke(100)
	state, err := sl.ToStateDict()
	if err != nil {
		t.Fatal(err)
	}
	raw, err := json.Marshal(state)
	if err != nil {
		t.Fatal(err)
	}
	var decoded StatusListStateDict
	if err := json.Unmarshal(raw, &decoded); err != nil {
		t.Fatal(err)
	}
	if decoded.NextIndex != state.NextIndex {
		t.Errorf("NextIndex mismatch after JSON round-trip")
	}
	if decoded.EncodedList != state.EncodedList {
		t.Errorf("EncodedList mismatch after JSON round-trip")
	}
}

func TestStateDictRejectsNil(t *testing.T) {
	if _, err := FromStateDict(nil); err == nil {
		t.Error("expected error for nil state")
	}
}

func TestStateDictRejectsOutOfRangeNextIndex(t *testing.T) {
	sl, _ := NewStatusList(testStatusURL, "", 0)
	state, _ := sl.ToStateDict()
	state.NextIndex = -1
	if _, err := FromStateDict(state); err == nil {
		t.Error("expected error for negative nextIndex")
	}
}

File: packages/sdk-ts/tests/status-list.test.ts
/**
 * Tests for the BitstringStatusList implementation (Specification §11.2).
 *
 * Mirrors `tests/test_status_list.py` and `go-sidecar/signer/status_list_test.go`.
 * Cross-language interop is verified against the canonical vector at
 * `test-vectors/bitstring-status-list/vector.json`.
 */

import * as fs from 'fs';
import * as path from 'path';

import { describe, expect, it } from 'vitest';

import {
  BITSTRING_STATUS_LIST_CREDENTIAL_TYPE,
  BITSTRING_STATUS_LIST_ENTRY_TYPE,
  BITSTRING_STATUS_LIST_SUBJECT_TYPE,
  DEFAULT_BITSTRING_LENGTH,
  MULTIBASE_BASE64URL_PREFIX,
  STATUS_PURPOSE_REVOCATION,
  STATUS_PURPOSE_SUSPENSION,
  StatusList,
  StatusListError,
  buildStatusListCredential,
  buildStatusListEntry,
  buildVouchCredential,
  verifyStatus,
} from '../src';

const STATUS_URL = 'https://issuer.example/status/1';
const ISSUER_DID = 'did:web:issuer.example';

const VECTOR_PATH = path.resolve(
  __dirname,
  '../../../test-vectors/bitstring-status-list/vector.json'
);

interface CanonicalVector {
  status_list_id: string;
  issuer_did: string;
  bitstring_length_bits: number;
  status_purpose: string;
  revoked_indices: number[];
  active_indices_sample: number[];
  expected_encoded_list: string;
  status_list_credential: Record<string, unknown>;
  sample_credential_status_revoked: Record<string, unknown>;
  sample_credential_status_active: Record<string, unknown>;
}

function loadVector(): CanonicalVector {
  return JSON.parse(fs.readFileSync(VECTOR_PATH, 'utf8')) as CanonicalVector;
}

// ---------------------------------------------------------------------------
// Construction & validation
// ---------------------------------------------------------------------------

describe('StatusList construction', () => {
  it('uses default length and purpose', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    expect(sl.length).toBe(DEFAULT_BITSTRING_LENGTH);
    expect(sl.statusPurpose).toBe(STATUS_PURPOSE_REVOCATION);
  });

  it('rejects bitstring shorter than protocol minimum', () => {
    expect(() => new StatusList({ statusListId: STATUS_URL, length: 1024 })).toThrow(
      StatusListError
    );
  });

  it('rejects non-multiple-of-8 length', () => {
    expect(
      () =>
        new StatusList({
          statusListId: STATUS_URL,
          length: DEFAULT_BITSTRING_LENGTH + 1,
        })
    ).toThrow(StatusListError);
  });

  it('rejects invalid status purpose', () => {
    expect(
      () =>
        new StatusList({
          statusListId: STATUS_URL,
          // @ts-expect-error: intentional bad value
          statusPurpose: 'bogus',
        })
    ).toThrow(StatusListError);
  });

  it('rejects empty id', () => {
    expect(() => new StatusList({ statusListId: '' })).toThrow(StatusListError);
  });

  it('default state has all bits clear', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    for (const idx of [0, 1, 7, 8, 100, 65535, DEFAULT_BITSTRING_LENGTH - 1]) {
      expect(sl.getStatus(idx)).toBe(false);
    }
  });
});

describe('Bit operations', () => {
  it('sets and gets a bit', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    sl.setStatus(42, true);
    expect(sl.getStatus(42)).toBe(true);
    expect(sl.getStatus(41)).toBe(false);
    expect(sl.getStatus(43)).toBe(false);
  });

  it('clears a bit after setting', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    sl.setStatus(42, true);
    sl.setStatus(42, false);
    expect(sl.getStatus(42)).toBe(false);
  });

  it('handles first and last bit', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    sl.revoke(0);
    sl.revoke(DEFAULT_BITSTRING_LENGTH - 1);
    expect(sl.isSet(0)).toBe(true);
    expect(sl.isSet(DEFAULT_BITSTRING_LENGTH - 1)).toBe(true);
    expect(sl.isSet(1)).toBe(false);
  });

  it('throws on out-of-range index', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    expect(() => sl.setStatus(-1)).toThrow(StatusListError);
    expect(() => sl.setStatus(DEFAULT_BITSTRING_LENGTH)).toThrow(StatusListError);
  });

  it('revoke / reinstate helpers work for suspension purpose', () => {
    const sl = new StatusList({
      statusListId: STATUS_URL,
      statusPurpose: STATUS_PURPOSE_SUSPENSION,
    });
    sl.revoke(7);
    expect(sl.isSet(7)).toBe(true);
    sl.reinstate(7);
    expect(sl.isSet(7)).toBe(false);
  });
});

describe('Allocation', () => {
  it('returns sequential indices', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    expect(sl.allocateIndex()).toBe(0);
    expect(sl.allocateIndex()).toBe(1);
    expect(sl.allocateIndex()).toBe(2);
  });
});

// ---------------------------------------------------------------------------
// Encoding
// ---------------------------------------------------------------------------

describe('Encoding', () => {
  it('uses multibase base64url prefix', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    expect(sl.encode().startsWith(MULTIBASE_BASE64URL_PREFIX)).toBe(true);
  });

  it('round-trip preserves all bits', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    const indices = [0, 1, 7, 8, 9, 16, 1023, 65535, DEFAULT_BITSTRING_LENGTH - 1];
    for (const idx of indices) {
      sl.revoke(idx);
    }
    const encoded = sl.encode();
    const decoded = StatusList.decode(encoded, STATUS_URL);
    for (const idx of indices) {
      expect(decoded.getStatus(idx)).toBe(true);
    }
    expect(decoded.getStatus(2)).toBe(false);
    expect(decoded.getStatus(50)).toBe(false);
  });

  it('rejects wrong multibase prefix', () => {
    expect(() => StatusList.decode('zSomethingBase58', STATUS_URL)).toThrow(StatusListError);
  });

  it('rejects corrupt payload', () => {
    expect(() =>
      StatusList.decode(`${MULTIBASE_BASE64URL_PREFIX}$$$invalid$$$`, STATUS_URL)
    ).toThrow(StatusListError);
  });
});

// ---------------------------------------------------------------------------
// Credential and entry builders
// ---------------------------------------------------------------------------

describe('buildStatusListCredential', () => {
  it('produces a standards-shaped credential', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    sl.revoke(7);
    const vc = buildStatusListCredential({
      issuerDid: ISSUER_DID,
      statusList: sl,
    });

    expect(vc['@context']).toEqual(['https://www.w3.org/ns/credentials/v2']);
    expect(vc.id).toBe(STATUS_URL);
    expect(vc.type).toContain('VerifiableCredential');
    expect(vc.type).toContain(BITSTRING_STATUS_LIST_CREDENTIAL_TYPE);
    expect(vc.issuer).toBe(ISSUER_DID);
    expect(vc.credentialSubject.id).toBe(`${STATUS_URL}#list`);
    expect(vc.credentialSubject.type).toBe(BITSTRING_STATUS_LIST_SUBJECT_TYPE);
    expect(vc.credentialSubject.statusPurpose).toBe(STATUS_PURPOSE_REVOCATION);
    expect(vc.credentialSubject.encodedList.startsWith(MULTIBASE_BASE64URL_PREFIX)).toBe(true);
  });
});

describe('buildStatusListEntry', () => {
  it('produces the standards-shaped entry', () => {
    const entry = buildStatusListEntry({
      statusListCredential: STATUS_URL,
      statusListIndex: 42,
    });
    expect(entry).toEqual({
      id: `${STATUS_URL}#42`,
      type: BITSTRING_STATUS_LIST_ENTRY_TYPE,
      statusPurpose: STATUS_PURPOSE_REVOCATION,
      statusListIndex: '42',
      statusListCredential: STATUS_URL,
    });
  });

  it('rejects negative index', () => {
    expect(() =>
      buildStatusListEntry({
        statusListCredential: STATUS_URL,
        statusListIndex: -1,
      })
    ).toThrow(StatusListError);
  });
});

// ---------------------------------------------------------------------------
// verifyStatus
// ---------------------------------------------------------------------------

describe('verifyStatus', () => {
  function buildPair(revoked: number[] = []) {
    const sl = new StatusList({ statusListId: STATUS_URL });
    for (const idx of revoked) {
      sl.revoke(idx);
    }
    return {
      sl,
      vc: buildStatusListCredential({ issuerDid: ISSUER_DID, statusList: sl }),
    };
  }

  it('returns false when bit is unset', () => {
    const { vc } = buildPair();
    const entry = buildStatusListEntry({
      statusListCredential: STATUS_URL,
      statusListIndex: 10,
    });
    expect(verifyStatus({ credentialStatus: entry, statusListCredential: vc })).toBe(false);
  });

  it('returns true when bit is set', () => {
    const { vc } = buildPair([10]);
    const entry = buildStatusListEntry({
      statusListCredential: STATUS_URL,
      statusListIndex: 10,
    });
    expect(verifyStatus({ credentialStatus: entry, statusListCredential: vc })).toBe(true);
  });

  it('throws on id mismatch', () => {
    const { vc } = buildPair();
    const entry = buildStatusListEntry({
      statusListCredential: 'https://other.example/status/9',
      statusListIndex: 0,
    });
    expect(() => verifyStatus({ credentialStatus: entry, statusListCredential: vc })).toThrow(
      StatusListError
    );
  });

  it('throws on purpose mismatch', () => {
    const sl = new StatusList({
      statusListId: STATUS_URL,
      statusPurpose: STATUS_PURPOSE_SUSPENSION,
    });
    const vc = buildStatusListCredential({ issuerDid: ISSUER_DID, statusList: sl });
    const entry = buildStatusListEntry({
      statusListCredential: STATUS_URL,
      statusListIndex: 0,
      statusPurpose: STATUS_PURPOSE_REVOCATION,
    });
    expect(() => verifyStatus({ credentialStatus: entry, statusListCredential: vc })).toThrow(
      StatusListError
    );
  });
});

// ---------------------------------------------------------------------------
// buildVouchCredential integration
// ---------------------------------------------------------------------------

describe('buildVouchCredential integration', () => {
  it('embeds credentialStatus when provided', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    const idx = sl.allocateIndex();
    const entry = buildStatusListEntry({
      statusListCredential: STATUS_URL,
      statusListIndex: idx,
    });

    const vc = buildVouchCredential({
      issuerDid: 'did:web:agent.example.com',
      intent: {
        action: 'POST',
        target: 'https://api.example.com/orders',
        resource: 'order:42',
      },
      credentialStatus: entry as unknown as Record<string, unknown>,
    });

    expect(vc.credentialStatus).toEqual(entry);
  });

  it('omits credentialStatus when not provided', () => {
    const vc = buildVouchCredential({
      issuerDid: 'did:web:agent.example.com',
      intent: {
        action: 'POST',
        target: 'https://api.example.com/orders',
        resource: 'order:42',
      },
    });
    expect(vc.credentialStatus).toBeUndefined();
  });
});

// ---------------------------------------------------------------------------
// Cross-language interop against the canonical test vector
// ---------------------------------------------------------------------------

describe('State dict persistence', () => {
  it('round-trips bits and cursor', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    const a = sl.allocateIndex();
    const b = sl.allocateIndex();
    const c = sl.allocateIndex();
    sl.revoke(b);

    const state = sl.toStateDict();
    const restored = StatusList.fromStateDict(state);

    expect(restored.statusListId).toBe(STATUS_URL);
    expect(restored.length).toBe(sl.length);
    expect(restored.getStatus(a)).toBe(false);
    expect(restored.getStatus(b)).toBe(true);
    expect(restored.getStatus(c)).toBe(false);
    expect(restored.allocateIndex()).toBe(c + 1);
  });

  it('state dict is JSON serializable', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    sl.revoke(100);
    const state = sl.toStateDict();
    const json = JSON.stringify(state);
    expect(JSON.parse(json)).toEqual(state);
  });

  it('rejects state dict missing required keys', () => {
    expect(() =>
      // @ts-expect-error: intentionally incomplete
      StatusList.fromStateDict({ statusListId: STATUS_URL })
    ).toThrow(StatusListError);
  });

  it('rejects out-of-range nextIndex', () => {
    const sl = new StatusList({ statusListId: STATUS_URL });
    const state = sl.toStateDict();
    state.nextIndex = -1;
    expect(() => StatusList.fromStateDict(state)).toThrow(StatusListError);
  });
});

describe('Cross-language interop', () => {
  it('encodes the canonical revoked-indices set to byte-identical output', () => {
    const v = loadVector();
    const sl = new StatusList({ statusListId: v.status_list_id });
    for (const idx of v.revoked_indices) {
      sl.revoke(idx);
    }
    expect(sl.encode()).toBe(v.expected_encoded_list);
  });

  it('decodes the canonical encodedList and reports correct bits', () => {
    const v = loadVector();
    const sl = StatusList.decode(
      v.expected_encoded_list,
      v.status_list_id,
      v.status_purpose as 'revocation'
    );
    for (const idx of v.revoked_indices) {
      expect(sl.getStatus(idx)).toBe(true);
    }
    for (const idx of v.active_indices_sample) {
      expect(sl.getStatus(idx)).toBe(false);
    }
  });

  it('verifyStatus matches the canonical credentials', () => {
    const v = loadVector();
    const revokedResult = verifyStatus({
      credentialStatus: v.sample_credential_status_revoked,
      statusListCredential: v.status_list_credential,
    });
    const activeResult = verifyStatus({
      credentialStatus: v.sample_credential_status_active,
      statusListCredential: v.status_list_credential,
    });
    expect(revokedResult).toBe(true);
    expect(activeResult).toBe(false);
  });
});
